src/api/initialDataTypes.js

```diff
@@ -19,7 +19,30 @@ import type {
 export type InitialDataBase = $ReadOnly<{|
   last_event_id: number,
   msg: string,
-  queue_id: number,
+  queue_id: string,
+
+  // The feature level and server version, below, are included
+  // unconditionally as of 3.0 (zulip/zulip@2c6313019), which is why we put
+  // them in `InitialDataBase`. To get them from servers older than that, we
+  // have to either
+  // - pass `zulip_version` in `event_types`, or
+  // - not specify `event_types`.
+  // We've been doing the latter for a long time, so we're fine.
+
```

Comment:
Huh. We *do* specify `fetch_event_types`, which controls most of what appears in the initial data. If it's `event_types` here, that seems like an anomaly.

I guess in that case we fixed it long ago by sending these unconditionally :shrug:

Reply:
I agree, it's odd for it to be `event_types` instead of `fetch_event_types`. But I think it is. From the quoted commit https://github.com/zulip/zulip/commit/2c6313019:

```
**Feature level 3**:

* `zulip_version` and `zulip_feature_level` are always returned
  in `POST /register`; previously they were only returned if `event_types`
  included `zulip_version`.
```

and `want`, as in the removed lines

```diff
-    if want('zulip_version'):
-        state['zulip_version'] = ZULIP_VERSION
-        state['zulip_feature_level'] = API_FEATURE_LEVEL
```

, is

```
    if event_types is None:
        # return True always
        want: Callable[[str], bool] = always_want
    else:
        want = set(event_types).__contains__
```

Reply:
But that's the same `want` that is used in the rest of that function too, for all the other parts of the initial data.

Looking at that function's callsite lower in that file, its `event_types` argument is computed like so:
```py
    if fetch_event_types is not None:
        event_types_set: Optional[Set[str]] = set(fetch_event_types)
    elif event_types is not None:
        event_types_set = set(event_types)
    else:
        event_types_set = None

    # …
        ret = fetch_initial_state_data(
            user_profile,
            event_types=event_types_set,
```

So `fetch_event_types` controls if it's present. I'm guessing that that name (a parameter to *that* function) corresponds to the same name in the HTTP request.

Reply:
Huh, interesting. Thanks for noticing that!